### DISCUSSION TITLE
Suggestion: Implement retention of scroll position of pages when Epub is scrollable

### DISCUSSION THREAD

**Discussion by rorodeveloper** [readium/kotlin-toolkit#572] [Ideas]
Description: Every page the user visited, the scroll position is saved and when the user navigates back to a page they already visited during the same session, it should still be focused on the last scroll position. Current behavior is that they navigate to the top when scrolling to a previous page and navigate to the top when navigating to a next page. This data will only last until EpubNavigatorFragment is destroyed. Probably add this setting as an additional EpubPreferences setting.

Reason: Swiping between pages can be really sensitive and when accidentally swiping when trying to highlight or copy something, the user would need to scroll back to the last position manually. For educational books, it is also a good feature to have so the user can easily navigate back and forth while comparing something between pages.

I have implemented this in a local project based on the sample app and can create a PR for this myself but I don't know how to get about this. How does this usually work for open source projects?

**Comment by qnga** [comment]
Hello,
I understand your concern about accidental swiping. I guess keeping the last position in some cases would make sense, but I'm not sure about the "same session" scope. For instance, that behaviour won't make sense if we jump to another location with `go`. @mickael-menu @HadrienGardeur ?

> I have implemented this in a local project based on the sample app and can create a PR for this myself but I don't know how to get about this. How does this usually work for open source projects?

Opening a PR is the right way. You can do it now so that we can have a look at your implementation, or later when the suitable behaviour for the toolkit is clearer.

**Comment by HadrienGardeur** [comment]
> Implement retention of scroll position of pages when Epub is scrollable

If by that you mean retention in the navigator, then I don't think that's necessarily a good idea.

But swiping between two resources should be treated like jumping to an item in the TOC or following a link. It should trigger an event indicating a discontinuity in the progression that contains a locator. Storing these locators would allow an app to implement a history forward/backward feature.

> Reason: Swiping between pages can be really sensitive and when accidentally swiping when trying to highlight or copy something, the user would need to scroll back to the last position manually. 

Swiping between resources is what we have currently in the toolkits but remains less than ideal.

In Readium Web, our preference is for using links at the bottom of the scroll instead. On mobile, this could be an overscroll since achieving continuous scrolling remains a challenge.

In the case of an accidental swipe for the time being, the best way to deal with them is to use the back feature and restore the previous position.

**Reply by mickael-menu** [reply]
That's the right approach IMHO. There's a big shortcoming with the current implementation in the fact that it's too easy to swipe a scrolling resource by mistake, loosing the position. Using a location history with a "go back" button could be a workaround until we find a better way to switch between scrolling resources.

**Reply by qnga** [reply]
I _think_ that swiping to the next resource doesn't trigger any event at the moment, at least in kotlin. There's no discontinuity, so it's not considered as a jump. It might be different in scroll mode but that would be weird.

**Reply by mickael-menu** [reply]
It probably should, because unless you are at the bottom or the top of the resource, it is a discontinuous move. In fact, we could check for the scroll progression to determine if it's a jump or continuous move.

What about just preventing the swipe if we're not at the to top or bottom of the resource?

**Reply by qnga** [reply]
Ok, I didn't think that Hadrien and you talked only about the case when you're not at the top or the bottom. Indeed, if you're not there's a discontinuity.

> What about just preventing the swipe if we're not at the to top or bottom of the resource?

That would make perfect sense to me. I'm not sure how difficult it is to do it in the implementation mess though.
